docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,323 @@
+you use a check that the type checker doesn't understand. In these cases, you
+can define a ``TypeIs`` or ``TypeGuard`` function to perform the check and allow type checkers
+to use it to narrow the type of a variable. Among the two, ``TypeIs`` usually
+has the more intuitive behavior, so we'll talk about it more; see
+:ref:`below <guide-type-narrowing-typeis-typeguard>` for a comparison.
+
+How to use ``TypeIs`` and ``TypeGuard``
+---------------------------------------
+
+A ``TypeIs`` function takes a single argument and is annotated as returning
+``TypeIs[T]``, where ``T`` is the type that you want to narrow to. The function
+must return ``True`` if the argument is of type ``T``, and ``False`` otherwise.
+The function can then be used in ``if`` checks, just like you would use ``isinstance()``.
+For example::
+
+    from typing import Literal, TypeIs
+
+    type Direction = Literal["N", "E", "S", "W"]
+
+    def is_direction(x: str) -> TypeIs[Direction]:
+        return x in {"N", "E", "S", "W"}
+
+    def maybe_direction(x: str) -> None:
+        if is_direction(x):
+            print(f"{x} is a cardinal direction")
+        else:
+            print(f"{x} is not a cardinal direction")
+
+A ``TypeGuard`` function looks similar and is used in the same way, but the
+type narrowing behavior is different, as dicussed in :ref:`the section below <guide-type-narrowing-typeis-typeguard>`.
+
+Depending on the version of Python you are running, you will be able to
+import ``TypeIs`` and ``TypeGuard`` either from the standard library :py:mod:`typing`
+module or from the third-party ``typing_extensions`` module:
+
+* ``TypeIs`` is in ``typing`` starting from Python 3.13 and in ``typing_extensions``
+  starting from version 4.10.0.
+* ``TypeGuard`` is in ``typing`` starting from Python 3.10 and in ``typing_extensions``
+  starting from version 3.10.0.0.
+
+
+Writing a correct ``TypeIs`` function
+-------------------------------------
+
+A ``TypeIs`` function allows you to override your type checker's type narrowing
+behavior. This is a powerful tool, but it can be dangerous because an incorrectly
+written ``TypeIs`` function can lead to unsound type checking, and type checkers
+cannot detect such errors.
+
+For a function returning ``TypeIs[T]`` to be correct, it must return ``True`` if and only if
+the argument is compatible with type ``T``, and ``False`` otherwise. If this condition is
+not met, the type checker may infer incorrect types.
+
+Below are some examples of correct and incorrect ``TypeIs`` functions::
+
+    from typing import TypeIs
+
+    # Correct
+    def is_int(x: object) -> TypeIs[int]:
+        return isinstance(x, int)
+
+    # Incorrect: does not return True for all ints
+    def is_positive_int(x: object) -> TypeIs[int]:
+        return isinstance(x, int) and x > 0
+
+    # Incorrect: returns True for some non-ints
+    def is_real_number(x: object) -> TypeIs[int]:
+        return isinstance(x, (int, float))
+
```

Comment:
These two examples would also be good to have in the `TypeIs` documentation at docs.python.org. Also, the `TypeIs` and `TypeGuard` docs could link to this document when it gets merged.